### PR TITLE
Refactor to remove the nxensite/nxdissite scripts

### PR DESCRIPTION
**Pull request by bmhughes**
# Description

- Refactor cookbook to remove nxensite/nxdissite
- Remove all configuration resources from nginx_install.
- Move all general configuration to nginx_config.
- Refactor nginx_site to remove use of scripts.
- Create nginx_service resource to manage nginx service.
- Remove passenger install option.

## Issues Resolved

- None

## Check List

- [x] All tests pass. See TESTING.md for details.
- [x] New functionality includes testing.
- [x] New functionality has been documented in the README if applicable.

### CODE REVIEW

**Comment by kitchen-porter** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big Pull Request.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a Table Flip.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -44,19 +44,17 @@ jobs:
           - 'amazonlinux-2'
           - 'centos-7'
           - 'centos-8'
-          - 'debian-8'
           - 'debian-9'
-          - 'fedora-31'
+          - 'debian-10'
+          - 'fedora-32'
```
Comment (ramereth) [inline .github/workflows/ci.yml:49]: This way you don't need to change it on new releases.
```suggestion
          - 'fedora-latest'
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,15 @@
 
 This file is used to list changes made in each version of the nginx cookbook.
 
+## 11.0.0 (2020-11-07)
```
Comment (ramereth) [inline CHANGELOG.md:5]: Let's not do a release as a part of the PR. We'll do that separately.
```suggestion
## Unreleased (2020-11-07)
```

File: documentation/nginx_config.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+[Back to resource list](../README.md#resources)
+
+## Actions
+
+- `:create`
+- `:delete`
+
+## Properties
+
+| Name                   | Type          | Default                          | Description                                                         | Allowed Values      |
```
Comment (ramereth) [inline documentation/nginx_config.md:12]: Is there any reason why we need the `Allowed Values` column if we're not using it?

File: documentation/nginx_service.md [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+- `:start`
+- `:stop`
+- `:restart`
+- `:reload`
+- `:enable`
+- `:disable`
+
+## Properties
+
+| Name                   | Type          | Default                          | Description                                                         | Allowed Values      |
```
Comment (ramereth) [inline documentation/nginx_service.md:16]: Is there any reason why we need the `Allowed Values` column if we're not using it?

File: documentation/nginx_site.md [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+[Back to resource list](../README.md#resources)
+
+## Actions
+
+- `:create` - Create an nginx site configuration file
+- `:delete` - Delete an nginx site configuration file
+
+## Properties
+
+| Name                   | Type          | Default                          | Description                                                         | Allowed Values      |
```
Comment (ramereth) [inline documentation/nginx_site.md:12]: Is there any reason why we need the `Allowed Values` column if we're not using it?

File: kitchen.dokken.yml [hunk 6]
```diff
@@ -2,66 +2,55 @@
...
       image: dokken/centos-7
       pid_one_command: /usr/lib/systemd/systemd
-
   - name: centos-8
     driver:
       image: dokken/centos-8
       pid_one_command: /usr/lib/systemd/systemd
-
-  - name: fedora-31
+  - name: fedora-32
```
Comment (ramereth) [inline kitchen.dokken.yml:38]: I recommend you use the following so we don't have to keep updating it on every release.
```suggestion
  - name: fedora-latest
```

File: kitchen.yml [hunk 7]
```diff
@@ -16,48 +21,52 @@ platforms:
   - name: amazonlinux-2
   - name: centos-7
   - name: centos-8
-  - name: debian-8
   - name: debian-9
-  - name: fedora-31
+  - name: debian-10
+  - name: fedora-32
```
Comment (ramereth) [inline kitchen.yml:26]:
```suggestion
  - name: fedora-latest
```

File: metadata.rb [hunk 8]
```diff
@@ -6,7 +6,7 @@
 chef_version     '>= 14'
 license          'Apache-2.0'
 description      'Installs and configures nginx'
-version          '10.6.0'
+version          '11.0.0'
```
Comment (ramereth) [inline metadata.rb:9]: Please don't bump cookbooks in PRs.

File: resources/install.rb [hunk 9]
```diff
@@ -184,10 +67,6 @@
         baseurl      repo_url
         gpgkey       repo_signing_key
       end
-      execute 'dnf -qy module disable nginx' do
-        only_if { node['platform_version'].to_i >= 8 && platform_family?('rhel') || platform_family?('fedora') }
-        not_if 'dnf module list nginx | grep -q "^nginx.*\[x\]"'
-      end
```
Comment (ramereth) [inline resources/install.rb:190]: Why is this being removed?

**Review by ramereth — CHANGES_REQUESTED** [review]
Can you please provide a little context for why these big changes are needed?

**Comment by bmhughes** [comment]
> 
> 
> Can you please provide a little context for why these big changes are needed?

I can try.

1. The version 10 cookbook is far less useful/flexible than version 9 was, I get the desire to remove all the recipes and attributes but it removed a **lot** of functionality as well. Especially for anything outside of a normal nginx http site, this doesn't put that back but goes further towards letting the implementing user do it themselves.

2. I'm failing to see anything but restrictions/added complexity with the en/dis scripts. As a site resource is only really a wrapped template resource, why not pass that flexibility through, right now an implementer is stuck to using that paradigm whether they wanted it or not. Afaik it was only ever an Ubuntu thing anyway.

3. There seems to be a general lack of structure, why does the install resource do a bucketful of configuration? Then do it again with the config resource.

4. The hard coded notifies in various resources, would it not be better to allow the wrapper to perform this so the implementing user can use their own process flow?

**Comment by ramereth** [comment]
@bmhughes can you please rebase and resolve the conflicts? Also fix the metadata validator issue.

**Review by ramereth — COMMENTED** [review]
I'm on my phone but I wanted to finish part of my review on here. I noticed that this code is not idempotent similar to how the dhcp cookbook is. I'd really to be able to fix that properly somehow. I'll try and take a closer look when I'm back on a laptop.

**Comment by bmhughes** [comment]
Hmm, thought I'd done the `fedora-latest` change already, not sure how that got missed. Done.

The service ruby_block is a chicken-and-the-egg problem, I can't really see a way to work around it and keep 100% **reliable** service (re)starts with a config test without being able to get service state beforehand on multiple platforms. Otherwise it's back to the notifications stuff that dhcpd used before which has it's own problems. Have a play around and see what you think, I ended up wondering if there was a way to get to the chef service providers to check via those and skip the ruby_block or use converge_by if you can get state..

**Comment by bmhughes** [comment]
@ramereth I've been doing some more testing and removing the logic from the `ruby_block` still works the same which I didn't expect. I'd have expected it to run into a compile/converge problem and be testing the old configuration during compile, but it doesn't. So I can remove the ruby_block to but I still find that strange.

**Comment by Xorima** [comment]
@ramereth can you re-review this when you get time please? Be great to get this in

**Comment by bmhughes** [comment]
> 
> 
> @ramereth can you re-review this when you get time please? Be great to get this in

I'm happily running with this now, but I'm also mindful of the religious war it seemed to spark at the time as well. That said the way there was configuration and service management happening in the install resource it wasn't exactly great as it was either.

On the whole I'm getting vibes of "always done it like that" so I'm happy to move forward with it 🤷🏻‍♂️.

**Review by ramereth — APPROVED** [review]

**Comment by kitchen-porter** [comment]
Released as: [11.0.0](https://github.com/sous-chefs/nginx/releases/tag/11.0.0)

**Pull request merged**